### PR TITLE
fix(goto): run overlay dismissal in an isolated world

### PR DESCRIPTION
**Pull request by Kikobeats**
## Summary

`dismissOverlays` ran in the page **main world**, so any page could observe it:

- `window.__browserlessDismiss` was a readable global.
- Every `querySelectorAll` the scan made went through the page's own `Document.prototype` / `Element.prototype`, so a hook installed by the page recorded the automation stack (`scan (<anonymous>)`, `Object.scan [as rescan] … pptr:evaluate`).

This moves overlay dismissal into a named isolated world in the **top frame only**, the same approach patchright uses for its helpers (a utility world that shares the DOM but not JavaScript globals).

- `setup`: registers `page.on('domcontentloaded')` once per page.
  - The event only fires for the main frame (measured: 2 per goto+reload with 5 child frames, versus 10 child-frame DOMContentLoaded lifecycle events).
  - Puppeteer re-binds it when a prerender activation swaps the page to a new CDP session.
  - The listener runs the dismissal in the main frame's isolated world. That dismissal is asynchronous, and goto's post-navigation `run` settles it before goto returns.
- `run`: `Page.createIsolatedWorld({ frameId: mainFrame, worldName })` + `Runtime.evaluate({ contextId })`.
  - Protocol errors meaning a stale context (`Cannot find context | navigated or closed | Execution context was destroyed`) are retried up to 3 times.
  - Other protocol errors and page exceptions are not retried.
- The idempotency guard only accepts an **own** `__browserlessDismiss` property, so an element with that id (named access on `window`) cannot switch dismissal off.

`Page.createIsolatedWorld` with an existing `worldName` returns the **same** context (measured: identical `executionContextId`, state preserved). So the DOMContentLoaded dismissal and goto's post-navigation `run` share one instance and one idempotency guard, and a dialog is never clicked twice.

### Why not `page.mainFrame().isolatedRealm()`

That realm is Puppeteer's own `__puppeteer_utility_world__<version>`. Raw CDP keeps one world name we own. The only internals used are `page._client()` (already used in `packages/goto/src/index.js:87`) and `frame._id`.

### Page observability

| Signal | origin/master | this PR |
|---|---|---|
| `__browserlessDismiss` visible to the page | **yes** | no |
| main-world stacks from dismiss (`scan`, `rescan … pptr:evaluate`) | **3** | 0 |

### Iframes

Interleaved runs, 9 samples per cell, medians:

| Page | master | first revision (`b15081f9`) | this PR |
|---|---|---|---|
| no iframes | 0 worlds · 3.20 MB · 530 ms | 1 · 3.43 MB · 529 ms | 1 · 3.42 MB · 530 ms |
| 50 same-origin iframes | 0 · 23.96 MB · 692 ms | **51 · 32.34 MB** · 692 ms | **1 · 24.14 MB** · 680 ms |
| 20 cross-origin iframes | 0 · 10.98 MB · 591 ms | **21 · 14.44 MB** · 593 ms | **1 · 11.16 MB** · 594 ms |

### Robustness

| Scenario | fc156df6 | this PR |
|---|---|---|
| navigations dismissed after a prerender activation (reload, raw `page.goto`, JS nav, 302 chain, cross-site) | **0 / 5** | 5 / 5 |
| page with `<div id="__browserlessDismiss">`: clicks through goto | **0** | 1 |
| `run` racing a `location.reload()` (master 0 / 120) | 0 / 120 rejected | 0 / 120 rejected |

### CDP calls per goto (adblock on)

| | total | dismiss's own calls |
|---|---|---|
| master | 32–33 | 2 |
| first revision (`b15081f9`) | 33–34 | 3 |
| this PR | 35–36 | 4 (2 on DOMContentLoaded + 2 in the post-navigation `run`) |

That is one more call per navigation than the first revision (median 0.1 ms per call). Caching the context ID to avoid it is unsafe: after a renderer process swap the ID can point at a different context, including the main world.

## Tests

`packages/goto/test/unit/dismiss/index.js`, 28 tests. Every mutation of the design dies (`mutate.sh`):

| Mutation | Killed by |
|---|---|
| whole file = fc156df6 | clobbered guard name; prerender activation |
| whole file = origin/master | 8 tests |
| no WeakSet (listener per goto) | once per document |
| retry every error | no retry for other protocol errors |
| no retry | navigation between CDP calls; stale context up to three attempts |
| retry page exceptions | no retry for page exceptions |
| listen to `framenavigated` | once per document |
| listen to every frame's DOMContentLoaded lifecycle | once per document; prerender activation |
| listener on `page._client()` | prerender activation |
| guard via named access | clobbered guard name |
| drop the DOMContentLoaded listener | 3 tests |
| `addScriptToEvaluateOnNewDocument` in all frames | top frame only |
| evaluate in the main world | main-world state; main-world queries |
| `rescan` no-op | re-scan |
| consent guard removed | consent dialog |
| reject guard removed | 4 CMP tests |

- **Prerender test:** it waits for the prerendered document to request an image instead of listening to Preload events, because enabling the Preload domain disables prerendering (`PrerenderingDisabledByDevTools`).
- **Re-scan test:** it settles the DOMContentLoaded instance before mounting its dialog, so only `rescan` can produce the click.
- **Top-frame guard:** the `window.self !== window.top` check is removed; `run` only evaluates in the main frame's world, so it could not be reached.

Local runs (macOS):

- `ava test/unit/dismiss/index.js`: **28/28 pass**.
- `pnpm --filter @browserless/goto test`: **94 pass, 3 fail**. The 3 failures are the WebGL evasion tests (`webgl vendor is not bot`, `webgl2 vendor is not bot`, `webgl renderer goes through ANGLE`). They fail identically with origin/master's `dismiss.js`: `getContext('webgl')` returns `null` on this machine with Chrome 152 and `--use-angle=gl`. They pass on Linux CI.
- `standard` on both files: clean.

## Measurable outcome

Dismiss leaves zero page-observable traces: no global, no main-world DOM calls. It removes one of the four `suspiciousGlobals` a detector sees on every browserless `goto`, costs one isolated world per page instead of one per frame, and keeps dismissing across prerender activations and pages that clobber its guard name.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0116zka7w2WKPSi3kqHyVAET

---

> [!NOTE]
> **Medium Risk**
> Changes how every adblock-enabled navigation injects and runs dismissal (CDP + async DOMContentLoaded), with broad test coverage but real behavior risk on edge-case dialogs and timing relative to `goto`.
> 
> **Overview**
> Moves overlay auto-dismiss out of the page **main world** into a named CDP **isolated world** (`browserless_dismiss`) on the **top frame only**, so pages cannot see `__browserlessDismiss` or hook `querySelector` stacks from dismissal.
> 
> **Injection lifecycle:** `setup` no longer uses `evaluateOnNewDocument`; it registers a one-time `domcontentloaded` listener that runs dismissal asynchronously. `run` evaluates the same script via `Page.createIsolatedWorld` + `Runtime.evaluate`, with up to three retries when the execution context is destroyed by navigation. The idempotency guard uses an **own** property check so a clobbering DOM id cannot disable dismissal.
> 
> **In-world robustness:** DOM work uses captured prototype methods (`.call`) to survive form **named-property** shadowing; acknowledge/close candidates are tried in order with failed clicks skipped; per-dialog and per-scan errors are swallowed so one hostile dialog does not stop the rest.
> 
> Unit tests are expanded to cover isolation, main-world query hooks, iframe/world count, prerender CDP session swap, stale-context retry behavior, SVG/non-HTML buttons, and autoconsent handoff.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bbca5663080aa5865d541fd0eca3fdec2b043332. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved dismissal of consent dialogs across page navigations, prerender activation, and repeated scans.
  - Prevented a problematic dialog or page element from interrupting dismissal of other dialogs.
  - Improved compatibility with pages that override standard browser methods or use conflicting element names.
  - Fixed handling of dialogs in nested frames and non-HTML elements.
  - Improved recovery from transient page-loading errors.
  - Prevented duplicate setup work during repeated page events.

### CODE REVIEW

**Comment by Kikobeats** [comment]
Addressed the second review round in fc156df6.

**Dismiss world in every frame (performance).** `Page.addScriptToEvaluateOnNewDocument({ worldName })` has no frame filter, so Chrome created a `browserless_dismiss` world in each iframe. `setup` now listens to `Page.domContentEventFired`, which only fires for the main frame, and runs the dismissal in the main frame's isolated world. Interleaved medians, 9 samples per cell:

| Page | master | b15081f9 | fc156df6 |
|---|---|---|---|
| no iframes | 0 worlds · 3.20 MB · 530 ms | 1 · 3.43 MB · 529 ms | 1 · 3.42 MB · 530 ms |
| 50 same-origin iframes | 0 · 23.96 MB · 692 ms | 51 · 32.34 MB · 692 ms | 1 · 24.14 MB · 680 ms |
| 20 cross-origin iframes | 0 · 10.98 MB · 591 ms | 21 · 14.44 MB · 593 ms | 1 · 11.16 MB · 594 ms |

CDP calls per goto with adblock on: master 32–33, b15081f9 33–34, fc156df6 35–36. Dismiss itself now sends 4 calls per navigation (2 on DOMContentLoaded, 2 in the post-navigation `run`), one more than b15081f9, at about 0.1 ms. Getting back to 3 would mean caching a context ID across calls, and after a renderer process swap that ID can resolve to another context, including the main world.

**Page exceptions were retried.** Only CDP protocol errors are retried now. A page exception whose message reads "Cannot find context" rejects after `createIsolatedWorld` + `evaluate` (2 sends instead of 6).

**Tests.** `creates the dismiss world only in the top frame` (3 srcdoc iframes, counts `Runtime.executionContextCreated` named `browserless_dismiss`) fails on b15081f9 with 4 worlds. `run does not retry page exceptions that look like a stale context` fails on b15081f9 with 6 sends.

Page close during `run` is unchanged: it rejects and goto swallows it, as on master.

**Comment by Kikobeats** [comment]
Addressed the third review round in 9a7245ac and 203bc877.

**F1: listener lost after prerender activation.** Activating a speculation-rules prerender swaps the page to a new CDP session. The `Page.domContentEventFired` listener stayed on the dead session, and the per-page guard blocked registering it again. `setup` now uses `page.on('domcontentloaded')`, which Puppeteer re-binds on the swap. Using `new.js prerender2` (activate, 3 gotos, then reload, raw `page.goto`, JS navigation, 302 chain, cross-site):

| | dismissed after activation |
|---|---|
| fc156df6 | 0 / 5 |
| this PR | 5 / 5 |

**F2: guard clobbered by a named element.** `<div id="__browserlessDismiss">` is a named property of `window` in the isolated world too. The guard now checks `Object.prototype.hasOwnProperty.call(window, '__browserlessDismiss')`. With `attack.js tamper` on the clobber-id page: fc156df6 0 clicks through `goto` (`run` alone rejects with `rescan is not a function`), this PR 1 click through `goto` and 1 through `run`.

**F3: tests now pin the design.** Every mutation dies, and so do both whole-file variants:

| Mutation | Result | Killed by |
|---|---|---|
| whole file = fc156df6 | killed | clobbered guard name; prerender activation |
| whole file = master | killed | 8 tests |
| no WeakSet (listener per goto) | killed | once per document |
| retry every error | killed | no retry for other protocol errors |
| no retry | killed | navigation between CDP calls; stale context up to three attempts |
| retry page exceptions | killed | no retry for page exceptions |
| listen to `framenavigated` | killed | once per document |
| listen to every frame's DOMContentLoaded lifecycle | killed | once per document; prerender activation |
| listener on `page._client()` (F1) | killed | prerender activation |
| guard via named access (F2) | killed | clobbered guard name |
| drop the DOMContentLoaded listener | killed | 3 tests |
| `addScriptToEvaluateOnNewDocument` in all frames | killed | top frame only |
| evaluate in the main world | killed | main-world state; main-world queries |
| `rescan` no-op | killed | re-scan |
| consent guard removed | killed | consent dialog |
| reject guard removed | killed | 4 CMP tests |

- **Top-frame guard:** the `window.self !== window.top` check is gone rather than tested. `run` only evaluates in the main frame's world, so nothing could reach it.
- **Re-scan test:** during mutation testing it survived the `rescan` no-op. The DOMContentLoaded dismissal could still be in flight when `page.goto` resolved, and its initial scan clicked the dialog. The test now settles the instance first; the mutation fails 3 / 3 and the unmutated test passes 3 / 3.
- **Prerender test:** it waits for the prerendered document to request an image instead of listening to Preload events, because enabling the Preload domain disables prerendering (`PrerenderingDisabledByDevTools`). It was stable in 5 / 5 probe runs: activated, session swapped, and the navigation outside goto dismissed.
- **Fallback test:** it now suppresses `page.on('domcontentloaded')`.

**F4.** `run` documents that the DOMContentLoaded dismissal is asynchronous, and that callers needing it settled await `run` afterwards, as goto does.

File: packages/goto/test/unit/dismiss/index.js [hunk 1]
```diff
@@ -293,10 +397,249 @@ test('re-scans on the post-navigation run for dialogs mounted after the initial
...
+    t,
+    STICKY + '<iframe srcdoc="<p>one</p>"></iframe><iframe srcdoc="<p>two</p>"></iframe>'
+  )
+
+  const run = browserless.withPage((page, goto) => async () => {
+    for (let navigation = 0; navigation < 3; navigation++) await goto(page, { url })
+    const client = page._client()
+    const send = client.send.bind(client)
+    let dismissWorldRequests = 0
+    client.send = (method, params, ...rest) => {
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/goto/test/unit/dismiss/index.js:678]: Keeping this as is on purpose. The test pins "one DOMContentLoaded dismissal per document", so it counts only a navigation that goto does not drive: the three gotos exist to register `setup` repeatedly, and the reload is the navigation measured. Counting the gotos too would add goto's post-navigation `run` (one extra `createIsolatedWorld` per goto) and any stale-context retry, so the expected number would track goto internals and navigation timing instead of the listener. It would not kill anything more either: the mutations this guards (a listener registered per goto, `framenavigated`, per-frame DOMContentLoaded lifecycle) already fail it with 3 requests on the reload instead of 1, verified with `mutate.sh`.

**Comment by Kikobeats** [comment]
Addressed the fourth review round in f0957e7c.

**F1: form named-access shadowing the DOM builtins.** A `<form>` dialog with `<input name="querySelectorAll">` (or `getClientRects`, `querySelector`) shadows that method on the form through HTMLFormElement named access, which reaches the isolated world too. Reading it off the element threw, the throw aborted the whole `scan` loop, and every later rescan threw again, so no dialog was dismissed. The scan now captures the builtins from the isolated world's prototypes once and calls them with the element as receiver, and wraps each dialog so one hostile node can't stop the scan or the rescans. Using the critic's `formdialog.js`, a hostile form followed by a normal sticky dialog:

| Case (sticky clicks) | 203bc877 | this PR |
|---|---|---|
| `<input name="getClientRects">`, at load / late | 0 / 0 (rejects) | 1 / 1 |
| `<input name="querySelector">`, at load / late | 0 / 0 (rejects) | 1 / 1 |
| `<input name="querySelectorAll">`, at load / late | 0 / 0 (rejects) | 1 / 1 |

The hostile form itself is never clicked (it has fields and its close button is inside the form).

**F2: untested stale-context patterns.** The retry test only covered `Cannot find context`, so narrowing the regex to just that pattern passed 28/28. There is now one retry unit test per message (`Cannot find context`, `Inspected target navigated or closed`, `Execution context was destroyed.`). Narrowing the regex to only `/Cannot find context/` now fails the other two.

**F5: two load-sensitive tests.** `does not touch a dialog without acknowledge buttons` waited a fixed 2s; it now awaits `dismiss.run(page)` so the scan has settled before asserting. `creates the dismiss world only in the top frame` counted worlds before the child frames were guaranteed attached; it now waits for all four frames first. The dismiss suite ran 5 consecutive clean times, 36/36 each.

New tests fail on 203bc877: the 6 form-shadow tests, and (under the F2 mutation) the two new stale-context patterns.

File: packages/goto/src/dismiss.js [hunk 2]
```diff
@@ -90,48 +106,52 @@ const dismissOverlays = () => {
...
 
     /* Single pass: defer the click so a reject button anywhere in the dialog
        still aborts it (leaving the opt-out to autoconsent), while the first
        acknowledge/close button becomes the candidate to click if none appears. */
     let candidate = null
     for (const button of buttons) {
       if (!isVisible(button) || button.disabled) continue
-      const text = normalize(button.innerText || button.value)
-      const label = normalize(button.getAttribute('aria-label'))
+      const text = normalize(readInnerText.call(button) || button.value)
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/goto/src/dismiss.js]: Fixed in 12f7ce76. Confirmed the regression first: a dialog with an `<svg role="button">` before a real OK button left the OK unclicked on f0957e7c (the innerText getter threw on the SVG, and the per-dialog catch skipped the dialog). `readText` now uses the `HTMLElement` innerText getter only for HTMLElements and falls back to `textContent` for anything else (SVG and other non-HTML elements are never `<form>`s, so the named-access shadowing the getter defends against does not apply to them). Added a regression test that fails on f0957e7c.

**Comment by Kikobeats** [comment]
Addressed the fifth review round in 703f109a and bbca5663.

**1. Non-HTML buttons (parity regression).**
- `readText` returns `''` for non-HTML elements again, matching what `element.innerText` always gave.
- Acknowledge and close candidates are now clicked in order, and the dialog is marked seen and counted only after `click.call` succeeds. If a click throws (an SVG close icon), the next candidate gets its turn.
- The `svg-button-title-reject` and `svg-dialog-foreignobject-cookietext` changes are reverted to master behaviour; neither was intended.

Corpus (`corpus.js`, 39 cases), PR vs master: 37 same, 2 different. Both differences are master bugs this PR removes, and both are tested:

| Case | master | this PR |
|---|---|---|
| `svg-button-arialabel-only` | run 1 with 0 clicks (phantom) | run 0 |
| `svg-three-close-dialogs-then-html` | rejects `candidate.click is not a function` | clicks the HTML dialog, run 1 |

New tests, each failing on 12f7ce76:
- the acknowledge button next to an SVG button titled "Close";
- next to an SVG button titled "Decline";
- next to an SVG button labelled "Close";
- next to a MathML `<mi role="button">x</mi>`, each asserting ok clicked and run 1;
- three unclickable SVG close dialogs followed by an HTML dialog (clicked, run 1).

**2. Tests now pin every defense.**
- Form dialogs whose `<button name=…>` shadows `getClientRects`, `querySelector` and `querySelectorAll`, at load and late.
- `<form role="button">` with a named `` shadowing `getClientRects`, `innerText`, `getAttribute`, `closest` (labelled Close, so `closest` is reached) and `click`.
- A throwing dialog followed by a normal one.

Mutation run (hook-based, `round5/mutate.js`): sanity 48/48, and all 21 mutations are killed:

| Mutation | Killed by |
|---|---|
| direct-rects | 3 form getClientRects tests |
| direct-qs / direct-qsa | form dialog querySelector / querySelectorAll, load and late |
| direct-getattr / direct-closest / direct-click / direct-innertext | the matching form button test |
| direct-all | 11 tests |
| no-trycatch | keeps scanning past a dialog whose button still throws |
| readtext-htmlonly | 5 non-HTML button tests |
| readtext-textcontent | SVG titled "Decline" |
| readtext-empty | 30 tests |
| count-before-click | SVG labelled "Close"; three SVG close dialogs |
| first-candidate-only | SVG labelled "Close" |
| stale-narrow | the two non-"Cannot find context" retry cases |
| no-weakset / client-listener / named-guard / rescan-noop / consent / reject guards | their existing tests |

`direct-doc` no longer applies (see 3).

**3. `scanDocument` removed.** Document named access isn't visible in isolated worlds, so it guarded nothing.

**Flake note.** In one of five full-file runs, `does not query the DOM from the page main world` failed on its click assertion. The test relied on the async DOMContentLoaded dismissal after a raw `page.goto`. It passed 15/15 in isolation with no listener errors, so it now awaits `run` before asserting, as documented on `run`.

**Verification.**
- Dismiss file: 48 tests, 5 consecutive clean runs after the test fix.
- `pnpm --filter @browserless/goto test`: 114 pass, 3 fail. The 3 are the macOS-only WebGL evasion tests, which fail identically with master's `dismiss.js` and pass on Linux CI.
- `standard`: clean.
- CI green on bbca5663.
